api/api/utils/strings.py

```diff
@@ -0,0 +1,53 @@
+import re
+from urllib.parse import quote, unquote
+
+
+def convert_grp(grp: str) -> str | None:
+    """
+    Convert a hex value into a character. Return None if the conversion results in
+    a character that cannot be used as a URI component.
+    """
+    try:
+        converted = chr(int(grp, 16))
+        # Decoded strings should be usable as URI components
+        quote(converted)
+        return converted
+    except UnicodeEncodeError:
+        return None
+
+
+def decode_data(data: str | None = "") -> str:
+    if not data:
+        return ""
+
+    regexes = [
+        r"\\(x)([\da-f]{2})",  # \\xe9 - é
+        r"\\(u)([\da-f]{4})",  # \\u00e9 - é
+        r"(u)([\da-f]{4})",  # u00e9 - é
+    ]
+
+    def replace_func(match):
+        """Replace the matched group with the converted character if possible, otherwise return the original string."""
+        prefix, grp = match.groups()
+        if converted := convert_grp(grp):
+            return converted
+        return f"{prefix}{grp}"
```

Comment:
Is it possible to use `encode().decode('unicode_escape')` instead of regexes:

```
>>> "Ol\\xe9".encode().decode('unicode_escape')
'Olé'

>>> "Ol\\u00e9".encode().decode('unicode_escape')
'Olé'
```

It doesn't work for unescaped unicode/hex markers though. Do we have a lot of circumstances where `u00e9` is unescaped? If the difference in complexity is between the two, I'd almost say it's not worth it, especially if we _must_ solve this at the serialization step (which I ask about elsewhere).

Reply:
The problem strings have _incorrectly_ modified encoded strings:
- instead of a single backslash before `x` or `u`, these strings have double backslash, i.e., `\\x` instead of `\x` and `\\u` instead of `\u`.
- some strings have _no_ backslash before `u`.

If we use `encode().decode`, we will break the correctly encoded strings, and not fix the strings that don't have a backslash.
```
>> incorrect = ["botu00f5es", "bot\\u00f5es", "bot\\xf5es", "botões"]
>> decoded = [x.encode().decode('unicode_escape') for x in incorrect]
["botu00f5es", "bot\\u00f5es", "bot\\xf5es", "botões"]
```
_Tags from https://api.openverse.engineering/v1/images/017ee061-3413-4c44-a610-077adf0635dd/_

So, we could divide this process into 2 steps: 
1. if the string has `r"\\(x)([\da-f]{2})"` or `r"\\(u)([\da-f]{4})"`: run the `encode.decode`
2. run the regex replacement with the no-backslash pattern:  `r"(u)([\da-f]{4})"`